File: python/clearings/src/clearings/flow.py
import asyncio
import inspect
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Generic, TypeVar, cast, overload

from ._control import Kind
from .values import ClearingsError, Codec, Value, copy_value

InputT = TypeVar("InputT")
OutputT = TypeVar("OutputT")


@dataclass(frozen=True)
class OperationContext:
    cancelled: asyncio.Event
    deadline: int


@dataclass(frozen=True)
class Operation(Generic[InputT, OutputT]):
    id: str
    version: str
    input: Codec[InputT]
    output: Codec[OutputT]
    execute: Callable[[InputT, OperationContext], Awaitable[OutputT]]
    effect: str = "read"


def operation(
    *,
    id: str,
    version: str,
    input: Codec[InputT],
    output: Codec[OutputT],
    execute: Callable[[InputT, OperationContext], Awaitable[OutputT]],
    effect: str = "read",
) -> Operation[InputT, OutputT]:
    if not id or not version or len(f"{id}@{version}".encode()) > 256 or effect != "read":
        raise ClearingsError("UNSUPPORTED")
    return Operation(id, version, input, output, execute)


@dataclass(frozen=True)
class Ref(Generic[OutputT]):
    owner: "Context"
    id: int


@dataclass
class Spec:
    kind: Kind
    deps: list[int]
    binding: str
    source: str
    value: Value = None
    execute: Callable[[list[Value], OperationContext], Value | Awaitable[Value]] | None = None
    operation: dict[str, str] | None = None


class Context:
    def __init__(self, name: str, reserve: Callable[[], None]) -> None:
        self.name = name
        self.nodes: list[Spec] = []
        self._reserve = reserve

    def _ref(self, spec: Spec) -> Ref[Any]:
        if not spec.source or len(spec.source.encode()) > 1024:
            raise ClearingsError("INVALID_PLAN")
        self._reserve()
        result: Ref[Any] = Ref(self, len(self.nodes))
        self.nodes.append(spec)
        return result

    def reference(self, ref: Ref[Any]) -> int:
        if not isinstance(ref, Ref) or ref.owner is not self or not 0 <= ref.id < len(self.nodes):
            raise ClearingsError("INVALID_PLAN")
        return ref.id

    def value(self, value: OutputT) -> Ref[OutputT]:
        return self._ref(
            Spec(
                "value",
                [],
                "",
                f"{self.name}:input:{len(self.nodes)}",
                copy_value(cast(Value, value)),
            )
        )

    def call(
        self, op: Operation[InputT, OutputT], input: InputT | Ref[InputT], source: str | None = None
    ) -> Ref[OutputT]:
        if op.effect != "read":
            raise ClearingsError("UNSUPPORTED")
        dep = input if isinstance(input, Ref) else self.value(input)
        label = source or f"{self.name}:{op.id}:{len(self.nodes)}"

        async def execute(args: list[Value], context: OperationContext) -> Value:
            argument = copy_value(args[0])
            if not op.input.accepts(argument):
                raise ClearingsError("INVALID_VALUE", label)
            result = copy_value(cast(Value, await op.execute(cast(InputT, argument), context)))
            if not op.output.accepts(result):
                raise ClearingsError("INVALID_VALUE", label)
            return result

        return self._ref(
            Spec(
                "call",
                [self.reference(dep)],
                f"{op.id}@{op.version}",
                label,
                execute=execute,
                operation={"id": op.id, "version": op.version},
            )
        )

    def transform(
        self, input: Ref[InputT], transform: Callable[[InputT], OutputT], source: str | None = None
    ) -> Ref[OutputT]:
        label = source or f"{self.name}:transform:{len(self.nodes)}"

        def execute(args: list[Value], context: OperationContext) -> Value:
            result = transform(cast(InputT, copy_value(args[0])))
            if inspect.isawaitable(result):
                if inspect.iscoroutine(result):
                    result.close()
                raise ClearingsError("UNSUPPORTED", label)
            return copy_value(cast(Value, result))

        return self._ref(
            Spec(
                "transform",
                [self.reference(input)],
                f"transform:{len(self.nodes)}",
                label,
                execute=execute,
            )
        )

    @overload
    def join(self, items: list[Ref[OutputT]]) -> Ref[list[OutputT]]: ...
    @overload
    def join(self, items: dict[str, Ref[Any]]) -> Ref[dict[str, Value]]: ...
    def join(self, items: list[Ref[Any]] | dict[str, Ref[Any]]) -> Ref[Any]:
        array = isinstance(items, list)
        refs = (
            cast(list[Ref[Any]], items)
            if array
            else list(cast(dict[str, Ref[Any]], items).values())
        )
        if len(refs) > 128:
            raise ClearingsError("CAPACITY")
        keys = [] if array else list(items)

        def execute(args: list[Value], context: OperationContext) -> Value:
            return args if array else dict(zip(cast(list[str], keys), args, strict=True))

        return self._ref(
            Spec(
                "join",
                [self.reference(r) for r in refs],
                "array" if array else "object",
                f"{self.name}:join:{len(self.nodes)}",
                execute=execute,
            )
        )

    def map(
        self, items: list[InputT], build: Callable[[InputT], Ref[OutputT]], max_items: int = 128
    ) -> Ref[list[OutputT]]:
        if (
            type(items) is not list
            or type(max_items) is not int
            or not 0 <= max_items <= 128
            or len(items) > max_items
        ):
            raise ClearingsError("CAPACITY")
        return self.join([build(item) for item in items])


@dataclass(frozen=True)
class Flow(Generic[InputT, OutputT]):
    name: str
    build: Callable[[Context, InputT], Ref[OutputT]]


def flow(name: str, build: Callable[[Context, InputT], Ref[OutputT]]) -> Flow[InputT, OutputT]:
    if not name or len(name.encode()) > 256:
        raise ClearingsError("INVALID_PLAN")
    return Flow(name, build)

File: packages/sdk/src/flow.ts
import { ClearingsError, copyValue, type Codec, type Value } from './values.js';
export interface OperationContext {
  readonly signal: AbortSignal;
  readonly deadline: number;
}
export interface Operation<I extends Value, O extends Value> {
  readonly id: string;
  readonly version: string;
  readonly input: Codec<I>;
  readonly output: Codec<O>;
  readonly effect: 'read';
  readonly execute: (input: I, context: OperationContext) => Promise<O>;
}
export function operation<I extends Value, O extends Value>(
  declaration: { id: string; version: string; input: Codec<I>; output: Codec<O>; effect?: 'read' },
  execute: (input: I, context: OperationContext) => Promise<O>,
): Operation<I, O> {
  if (
    Object.keys(declaration).some(
      (k) => !['id', 'version', 'input', 'output', 'effect'].includes(k),
    )
  )
    throw new ClearingsError('UNSUPPORTED');
  if (
    !declaration.id ||
    !declaration.version ||
    `${declaration.id}@${declaration.version}`.length > 256 ||
    (declaration.effect && declaration.effect !== 'read')
  )
    throw new ClearingsError('UNSUPPORTED');
  return Object.freeze({ ...declaration, effect: 'read', execute });
}
export class Ref<T extends Value> {
  declare private readonly resultType: T;
  constructor(
    readonly owner: Context,
    readonly id: number,
  ) {}
}
export interface Spec {
  kind: 'value' | 'call' | 'transform' | 'join';
  deps: number[];
  binding: string;
  source: string;
  value?: Value;
  execute?: (inputs: Value[], context: OperationContext) => Value | Promise<Value>;
  operation?: { id: string; version: string };
}
type Resolved<T> = T extends Ref<infer V> ? V : never;
export class Context {
  readonly nodes: Spec[] = [];
  constructor(
    readonly name: string,
    private readonly reserve: () => void,
  ) {}
  private ref<T extends Value>(spec: Spec): Ref<T> {
    if (spec.source.length > 1024 || !spec.source) throw new ClearingsError('INVALID_PLAN');
    this.reserve();
    const id = this.nodes.length;
    this.nodes.push(spec);
    return new Ref<T>(this, id);
  }
  reference<T extends Value>(ref: Ref<T>): number {
    if (!(ref instanceof Ref) || ref.owner !== this || !this.nodes[ref.id])
      throw new ClearingsError('INVALID_PLAN');
    return ref.id;
  }
  value<T extends Value>(value: T): Ref<T> {
    return this.ref({
      kind: 'value',
      deps: [],
      binding: '',
      source: `${this.name}:input:${this.nodes.length}`,
      value: copyValue(value),
    });
  }
  call<I extends Value, O extends Value>(
    op: Operation<I, O>,
    input: I | Ref<I>,
    source = `${this.name}:${op.id}:${this.nodes.length}`,
  ): Ref<O> {
    if (op.effect !== 'read') throw new ClearingsError('UNSUPPORTED');
    const dep = input instanceof Ref ? input : this.value(input);
    return this.ref({
      kind: 'call',
      deps: [this.reference(dep)],
      binding: `${op.id}@${op.version}`,
      source,
      operation: { id: op.id, version: op.version },
      execute: async (args, context) => {
        const argument = copyValue(args[0]!);
        if (!op.input.accepts(argument)) throw new ClearingsError('INVALID_VALUE', source);
        const result = copyValue(await op.execute(argument, context));
        if (!op.output.accepts(result)) throw new ClearingsError('INVALID_VALUE', source);
        return result;
      },
    });
  }
  transform<I extends Value, O extends Value>(
    input: Ref<I>,
    transform: (value: I) => O,
    source = `${this.name}:transform:${this.nodes.length}`,
  ): Ref<O> {
    return this.ref({
      kind: 'transform',
      deps: [this.reference(input)],
      binding: `transform:${this.nodes.length}`,
      source,
      execute: (args) => {
        const result = transform(copyValue(args[0]!) as I);
        if (result instanceof Promise) {
          void result.catch(() => {});
          throw new ClearingsError('UNSUPPORTED', source);
        }
        return copyValue(result);
      },
    });
  }
  join<T extends readonly Ref<Value>[]>(items: T): Ref<Resolved<T[number]>[]>;
  join<T extends Record<string, Ref<Value>>>(items: T): Ref<{ [K in keyof T]: Resolved<T[K]> }>;
  join(items: readonly Ref<Value>[] | Record<string, Ref<Value>>): Ref<Value> {
    const array = Array.isArray(items);
    const refs = array ? items : Object.values(items);
    if (refs.length > 128) throw new ClearingsError('CAPACITY');
    const keys = array ? [] : Object.keys(items);
    return this.ref({
      kind: 'join',
      deps: refs.map((r) => this.reference(r)),
      binding: array ? 'array' : 'object',
      source: `${this.name}:join:${this.nodes.length}`,
      execute: (args) => (array ? args : Object.fromEntries(keys.map((key, i) => [key, args[i]!]))),
    });
  }
  map<I extends Value, O extends Value>(
    items: readonly I[],
    build: (item: I, index: number) => Ref<O>,
    maxItems = 128,
  ): Ref<O[]> {
    if (
      !Array.isArray(items) ||
      !Number.isInteger(maxItems) ||
      maxItems < 0 ||
      maxItems > 128 ||
      items.length > maxItems
    )
      throw new ClearingsError('CAPACITY');
    return this.join(items.map((item, index) => build(item, index))) as Ref<O[]>;
  }
}
export interface Flow<I extends Value, O extends Value> {
  readonly name: string;
  readonly build: (context: Context, input: I) => Ref<O>;
}
export function flow<I extends Value, O extends Value>(
  name: string,
  build: (context: Context, input: I) => Ref<O>,
): Flow<I, O> {
  if (!name || name.length > 256) throw new ClearingsError('INVALID_PLAN');
  return Object.freeze({ name, build });
}
